google/gax/api_callable.py

```diff
@@ -73,22 +87,54 @@ def _retryable(a_func, retry):
         A function that will retry on exception.
     """
 
-    max_attempts = int(retry.backoff_settings.total_timeout_millis /
-                       retry.backoff_settings.initial_rpc_timeout_millis)
+    delay_mult = retry.backoff_settings.retry_delay_multiplier
+    max_delay = retry.backoff_settings.max_retry_delay_millis
+    timeout_mult = retry.backoff_settings.rpc_timeout_multiplier
+    max_timeout = retry.backoff_settings.max_rpc_timeout_millis
+    total_timeout = retry.backoff_settings.total_timeout_millis
 
     def inner(*args, **kwargs):
-        "Retries a_func upto max_attempt times"
-        attempt_count = 0
-        while 1:
+        """Equivalent to ``a_func``, but retries upon transient failure.
+
+        Retrying is done through an exponential backoff algorithm configured
+        by the options in ``retry``.
+        """
+        delay = retry.backoff_settings.initial_retry_delay_millis
+        timeout = retry.backoff_settings.initial_rpc_timeout_millis
+        result = OPTION_INHERIT
+        timer = TIMER_FACTORY(total_timeout / _MILLIS_PER_SECOND,
+                              lambda: None)
```

Comment:
Restructure this

- It is confusing as there is a timed loop *and* a timer; is the timer necessary ?
- It seems like the loop is only using the timer to keep track of time
  * i.e the timer callback does not do anything 
  * this is expensive; the default implementation of TIMER_FACTORY runs the timer on a separate thread
-  As it stands, this could probably be re-written using calls to time.time() to schedule the calls without the timer